### PR TITLE
python311Packages.tinygrad: 0.8.0 -> 0.9.0

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

Changelog: https://github.com/tinygrad/tinygrad/releases/tag/v0.9.0

Notable change: `gpuctypes` has been included directly in the code base. So we have to port our GPU-dependent magic in here...

cc @SomeoneSerge 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/tinygrad/default.nix [hunk 1]
```diff
@@ -96,11 +144,57 @@ buildPythonPackage rec {
...
+  # Require GPU access to run (not available in the sandbox)
+  pytestFlagsArray =
+    lib.optionals (!testCudaRuntime) [
+      "-k"
+      "'not TestCUDADevice'"
+    ]
+    ++ lib.optionals (!testRocmRuntime) [
+      "-k"
+      "'not TestHIPDevice'"
+    ]
```
Comment (natsukium) [inline pkgs/development/python-modules/tinygrad/default.nix:167]: Can we add these items to `disabledTests`?
e.g.

```nix
disabledTests = [
  ...
] ++ lib.optionals (!testCudaRuntime) [
  "not TestCUDADevice"
] ++ ...
```

Comment (GaetanLepage) [reply pkgs/development/python-modules/tinygrad/default.nix]: In the end, those test do not exist at all.

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 316931` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 316931` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 316931` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 316931` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 316931` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 316931` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
> Result of `nixpkgs-review pr 316931 --extra-nixpkgs-config '{ rocmSupport = true; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

Could you share the logs for this ? I don't have AMD hardware to test it on...

**Comment by pbsds** [comment]
https://hydra.nixos.org/build/261404072

**Comment by GaetanLepage** [comment]
> https://hydra.nixos.org/build/261404072

So this is a torch failure then and not a tinygrad regression, right ?

**Comment by pbsds** [comment]
> So this is a torch failure then and not a tinygrad regression, right ?

correct, making testing tinygrad on rocm impossible

**Comment by pbsds** [comment]
I'm getting a bunch of `sqlite3.OperationalError: database is locked` test failures for the cpu version, could be because of xdist and the fact that this host has 96 cores,

**Pull request merged**